### PR TITLE
reworking colander schema validation [WIP]

### PR DESCRIPTION
**Pull request by tisdall**
Okay, due to all the issues and comments surrounding colander validation I started working on a replacement.  Basically a lot of things were being done in Cornice when they could have easily been left to Colander (like iterating over child elements).  The schema.py file is now 62 lines where the old one was 180.  The 4 main components (querystring, headers, body, and path) are grouped into one structure and passed through a single schema for validation tests.  Each of those 4 components, if validated, are moved to `request.validated`.  Any errors are put in `request.errors`.

This is obviously a huge change from the previous method, so I wanted to get some input before I complete all aspects of this PR (please ignore the failing tests for the moment).

I modified a few existing tests to kind of demonstrate how the system works on validating the body but obviously more are needed to test validating the other 3 types.

This deals with the following issues: #310, #211, #190, #327, #193, #249 

Perhaps we could support #173 by calling unflatten in the validator.

### CODE REVIEW

**Comment by tisdall** [comment]
BTW, don't bother looking at the diff's for the two files as they're pretty drastic changes and it's probably easier to just read the whole file...

**Comment by tisdall** [comment]
As per the discussion on #329, I've updated this PR to not call colander's `serialize()` any more and just manually create the cstruct.  This way the validation schema only has to take the `deserialize()` process into account.

**Comment by tisdall** [comment]
I'd like to hear back from someone with the ability to merge this PR before continuing work on it...  I don't want to spend more time on it if it won't be used.

**Comment by leplatrem** [comment]
Hi @tisdall,

Thanks a lot for tackling this! It was extremely necessary!

You can definitely go on with making the tests pass :) I'll take some time to review the pull-requests about schemas (like #337), and give as much feedback as I can :)

Thanks (*again*) :)

File: cornice/schemas.py [hunk 1]
```diff
@@ -1,180 +1,78 @@
...
+            result = {}
+            for k in val.keys():
+                result[k] = simple_cstruct_serialize(val[k])
+            return result
+        except (TypeError, AttributeError):
+            try:
+                # try iterable interpretation
+                result = []
+                for k in val:
+                    result.append(simple_cstruct_serialize(k))
```
Comment (leplatrem) [inline cornice/schemas.py]: When looking at this: https://travis-ci.org/mozilla-services/cornice/jobs/76151945#L525 recursivity seems infinite!

File: cornice/schemas.py [hunk 2]
```diff
@@ -1,180 +1,78 @@
...
+    class StrictMappingSchema(colander.MappingSchema):
+        @staticmethod
+        def schema_type():
+            return colander.Mapping(unknown='raise')
+
+    class CorniceSchema(colander.MappingSchema):
+        querystring = StrictMappingSchema()
+        headers = StrictMappingSchema()
+        body = StrictMappingSchema()
+        path = StrictMappingSchema()
```
Comment (leplatrem) [inline cornice/schemas.py]: For retro-compatibility, we could add a ``from_colander`` method (?)

```python
    @classmethod
    def from_colander(klass, colander_schema, **kwargs):
        # XXX: ...written on a hungry stomache:
        class BodySchema(CorniceSchema):
            body = colander_schema
        return BodySchema(**kwargs)
```

Thoughts ?

File: cornice/schemas.py (hunk 1 above)
Comment (tisdall) [inline cornice/schemas.py]: yep.  it seems I didn't properly test `simple_cstruct_serialize()`.  It's because strings are iterable.  I'll patch this.

File: cornice/schemas.py (hunk 2 above)
Comment (tisdall) [inline cornice/schemas.py]: This PR works totally different then the past system so maybe it'd be best to actually change the name so there's no longer a CorniceSchema.  Or, CorniceSchema is some sort of compatibility class that's deprecated that helps people transition to the new system.

Any ideas for a different name?  Maybe it should be "ColanderSchema" since we're talking about adding alternate schema libraries and this class is already within the `cornice` namespace?

Comment (tisdall) [inline cornice/schemas.py]: Is "CorniceColanderSchema" too wordy?

Comment (tisdall) [inline cornice/schemas.py]: Another issue with the old system is everything ends up in `request.validated` so, for example, if you have a "body" element and a "querystring" with the same name they'll conflict.  The new one ends up with `request.validated['body']` and `request.validated['querystring']`.  I wrote a `from_colander()` method, but the result in request.validated will be completely different then before.

I don't think there's a way to support a smooth transition from the old...  So, I think it's best to fail early and hard so people are forced to face the change directly.

**Comment by delijati** [comment]
This is perfect. I'm currently working on #285 swagger integration and it would make the whole process of generating a **consistent** swagger schema much easier. Ony one schema for ever location. So only for body we could allow nested schemes.

The current validation process is too complicated a simple `schema.deserialize(data)` would be enough, if we got rid of the `location` and use this pull request (seperate schema for body, path, querystring, header).

**Comment by delijati** [comment]
This would make it possible to use https://github.com/stefanofontanelli/ColanderAlchemy ;)

**Comment by tisdall** [comment]
Hey guys...  I'm not really working on the project where I used cornice any more and other responsibilities  have come up that make it even more unlikely that I'll be able to finish this.  Hopefully someone else can pickup where I left off.

**Comment by leplatrem** [comment]
See how we resolved this https://github.com/Cornices/cornice/blob/master/docs/source/upgrading.rst#colander-validation

**Comment by delijati** [comment]
Nicely solved :)
